### ISSUE TITLE
Memory Leak in native Tracker

### ISSUE THREAD

**Issue by LeStahL** [vsariola/sointu#161]
After leaving the native tracker running for several hours, sointu crashes with the following output:

[sointu_error.log](https://github.com/user-attachments/files/17265231/sointu_error.log)

```
runtime: VirtualAlloc of 17955790848 bytes failed with errno=1455
fatal error: out of memory
```

I use the current master to reproduce.

**Comment by vsariola** [comment]
Was this with the native VST or the native stand-alone app?

**Comment by LeStahL** [comment]
native stand-alone app

**Comment by vsariola** [comment]
The crash is clearly from allocating a new synth object, which is ~16 meg thing with all the 64 delay buffers (65536 float32s each); for reasons we need to allocate all of them always. But we anyway allocate a new synth only when changing something in the patch so I don't think this is causing the leak itself. So the log isn't revealing too much; just that for whatever reason, there is no 16 meg available memory, because of leak...

Trying to memprofile to see if I can spot where all the allocations are done. I'm not convinced this is an issue with only the native version - the native version has cgo calls, but there is no C-style heap allocations (malloc etc.) taking place in those calls. The vst2 library has some, but this wasn't vst.

**Comment by vsariola** [comment]
Oh, what platform are you running? I realized the memory leak could be in oto, the library we use to output sound in a cross-platform manner. Currently sointu is still at oto v0.6.0 and current latest is v3.3.0... The API has also changed so upgrading was not that easy but I think I got it figured out. So maybe upgrading oto to v0.3.0 could solve the issue?

**Comment by LeStahL** [comment]
windows 10!

**Comment by vsariola** [comment]
I have trouble reproducing.

Did you run the code with the optimizations disabled i.e. -gcflags=all="-N -l". This happens when e.g. debugging from Visual Studio Code, as it needs to disable the optimizations so that each line has 1-1 correspondence to the source code. One of the consequences of the optimization disabling is that `arr = append(arr,make([]slice,n))` actually makes that slice, whereas with optimizations, it's clever enough that we are essentially just adding empty data to the end of arr and does not allocate it. This kind of allocations could perhaps lead to e.g. severe memory fragmentation to crash the app... But this is a bit of a stretch theory.

**Comment by LeStahL** [comment]
I can provide more details on what was unusual about the setup:
* I had several sync tracks
* I used gm.dls
* I had ~15 instruments
Will create a minimal reproducing patch; might take a while tho because of the inherent long reproduction times :)

**Comment by vsariola** [comment]
Any chance of getting more info on this? I have upgraded oto, don't know if that solves anything. If not, I have a bit trouble seeing how this could be specific to the -native tracker. There certainly could be memory leak in both types of tracker though :)

**Comment by LeStahL** [comment]
Sry, lazy me. OMW!

**Comment by LeStahL** [comment]
I tried to reproduce with the native tracker from the current master and did not see this again (while trying many different things over the course of 6 hours that I did when I first observed it). Neither with the exact track and on the hardware I saw this on first, nor on 3 different setups I have access to. I can not even make the tracker executable alloc more than 150 MB of RAM.

I also ran the tracker with pprof enabled, which was not able to identify any memory problems whatsoever.

With the version from back then (https://github.com/vsariola/sointu/commit/8fd2df19a16d16a7846a7ec02fb33e0422339b43), I can reproduce it by leaving the native tracker running for at least 3-4 hours with the screen locked.

I assume the oto update fixed it. Will leave the test running for another 8 hours and close if I don't see the problem again.
